Tests/UnitTests/Networking/ETagManagerTests.swift

```diff
@@ -212,17 +286,214 @@ class ETagManagerTests: TestCase {
+                                       statusCode: .notModified),
+            request: request,
+            retried: true
+        )
         expect(response).toNot(beNil())
         expect(response?.statusCode) == .notModified
         expect(response?.body) == actualResponse
+        expect(response?.validationResult) == .notRequested
+    }
+
+    func testETagHeaderIsNotFoundIfItsMissingResponseValidation() {
+        let eTag = "the_etag"
+        let url = self.urlForTest()
+        let request = URLRequest(url: url)
+        let cacheKey = url.absoluteString
+
+        let actualResponse = "response".asData
+
+        self.mockUserDefaults.mockValues[cacheKey] = """
+        {
+        "e_tag": "\(eTag)",
+        "status_code": 200,
+        "data": "\(actualResponse.asFetchToken)"
+        }
+        """.asData
+
+        let response = self.eTagManager.eTagHeader(for: request)
+        expect(response[ETagManager.eTagHeaderName]).to(beEmpty())
+    }
+
+    func testETagHeaderIsIgnoredIfValidationFailed() {
+        let eTag = "the_etag"
+        let url = self.urlForTest()
+        let request = URLRequest(url: url)
+        let cacheKey = url.absoluteString
+
+        let actualResponse = "response".asData
+
+        self.mockUserDefaults.mockValues[cacheKey] = ETagManager.Response(
+            eTag: eTag,
+            statusCode: .success,
+            data: actualResponse,
+            validationResult: .failedValidation
+        ).asData()
+
+        let response = self.eTagManager.eTagHeader(for: request)
+        expect(response[ETagManager.eTagHeaderName]).to(beEmpty())
+    }
+
+    func testETagHeaderIsReturnedIfValidationSucceded() {
+        let eTag = "the_etag"
+        let url = self.urlForTest()
+        let request = URLRequest(url: url)
+        let cacheKey = url.absoluteString
+
+        let actualResponse = "response".asData
+
+        self.mockUserDefaults.mockValues[cacheKey] = ETagManager.Response(
+            eTag: eTag,
+            statusCode: .success,
+            data: actualResponse,
+            validationResult: .validated
+        ).asData()
+
+        let response = self.eTagManager.eTagHeader(for: request)
+        expect(response[ETagManager.eTagHeaderName]) == eTag
+    }
+
+    func testETagHeaderIsIgnoredIfValidationWasNotRequested() {
+        let eTag = "the_etag"
+        let url = self.urlForTest()
+        let request = URLRequest(url: url)
+        let cacheKey = url.absoluteString
+
+        let actualResponse = "response".asData
+
+        self.mockUserDefaults.mockValues[cacheKey] = ETagManager.Response(
+            eTag: eTag,
+            statusCode: .success,
+            data: actualResponse,
+            validationResult: .notRequested
+        ).asData()
+
+        let response = self.eTagManager.eTagHeader(for: request)
+        expect(response[ETagManager.eTagHeaderName]) == eTag
+    }
+
+    func testCachedResponseWithNoValidationResultIsIgnored() {
```

Comment:
This tests backwards compatibility with old responses. We handle that as a cache miss.